File: leetcode/MayLeetCodePy/singleNonDuplicate.py
def singleNonDuplicate(nums: [int]) -> int:
    # print(nums)
    start = 0
    end = len(nums)-1
    # # print(len(nums[0:3+1]))
    # # print(len(nums[3:6]))
    while start <= end:
        mid = (start+end)//2
        # print("start:", start)
        # print("mid:", mid)
        # print("end:", end)
        if (mid < end and nums[mid] != nums[mid-1] and  nums[mid] != nums[mid+1]) or start == end:
            return nums[mid]
        elif nums[mid] == nums[mid-1]: #equal with left part
            # print('len from left',nums[start:mid+1])
            # print('len from right',nums[mid+1:end+1])
            if len(nums[start:mid+1])%2 !=0:
                end = mid
            elif len(nums[mid+1:end+1])%2 !=0:
                start = mid+1
        elif nums[mid] == nums[mid+1]:  #equal with right part
            # print('len from left',nums[start:mid])
            # print('len from right',nums[mid:end+1])
            if len(nums[start:mid])%2 !=0:
                end = mid -1
            elif len(nums[mid:end+1])%2 !=0:
                start = mid
    # print(set(nums))
    # print(set(nums)*2)
    # print(sum((set(nums))*2))
    # print(sum(nums))
    return (sum(set(nums))*2 - sum(nums))

print(singleNonDuplicate([3,3,7,7,10,11,11]))

File: leetcode/maychallenge/singleNonDuplicate.go
package main

import "fmt"

func main() {
	nums := []int{1, 1, 2, 3, 3, 4, 4, 8, 8}
	fmt.Println(singleNonDuplicate(nums))
}

// func singleNonDuplicate(nums []int) int {
// 	fmt.Println(nums[len(nums)-1])
// 	list := make([]int, nums[len(nums)-1]+1)
// 	for _, l := range nums {
// 		list[l]++
// 	}
// 	for i, l := range list {
// 		if l == 1 {
// 			return i
// 		}
// 	}
// 	// fmt.Println(list)
// 	return 0
// }

func singleNonDuplicate(nums []int) int {
	fmt.Println(nums)
	s, e := 0, len(nums)-1
	for s < e {
		m := s + (e-s)/2
		// doesn't match left or right element, this is the single element
		if (m > 0 && nums[m] != nums[m-1]) && (m < len(nums)-1 && nums[m] != nums[m+1]) {
			return nums[m]
		} else {
			// even index
			if m%2 == 0 {
				// matches left
				if m > 0 && nums[m] == nums[m-1] {
					e = m - 2
					// matches right
				} else {
					s = m + 2
				}
				// odd index
			} else {
				// matches left
				if m > 0 && nums[m] == nums[m-1] {
					s = m + 1
					// matches right
				} else {
					e = m - 1
				}
			}
		}
	}
	return nums[s]
}
